sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CollationTypeCasts.scala

```diff
@@ -78,6 +78,11 @@ object CollationTypeCasts extends TypeCoercionRule {
       }
       raiseError.withNewChildren(Seq(raiseError.errorClass, newErrorParams))
 
+    case framelessOffsetWindow @ (_: Lag | _: Lead) =>
```

Comment:
does the normal type coercion rules handle lead/lag as well?

Reply:
Apart from implicit casting for inputs no. But this change introduces only implicit casting for collations. The problem is that we have 
```
  override def inputTypes: Seq[AbstractDataType] =
    Seq(AnyDataType, IntegerType, TypeCollection(input.dataType, NullType))
```

Because of this we need to make sure we know what behaviour is expected. I would say as last parameter is default value, we would expect this not to fail when we pass a literal with default collation i.e. we would need to make sure that literal is cast to the proper collation for the `inputTypeCheck` not to fail.

Reply:
so we only want to allow implicit cast for lead/lag regarding string collations?

Reply:
Yes exactly.